File: kotlin/src/test/kotlin/com/solana/mpp/client/PaymentChannelsTest.kt
package com.solana.mpp.client

import com.solana.mpp.crypto.Blake3
import com.solana.mpp.crypto.Programs
import com.solana.mpp.crypto.PublicKey

import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * Byte-for-byte parity oracles for the payment-channels helpers. Golden
 * values mirror the Rust spine
 * (`rust/crates/mpp/src/program/payment_channels.rs::tests`) and the Go
 * port (`go/protocols/mpp/program/payment_channels_test.go`).
 */
class PaymentChannelsTest {
    private fun pk(byte: Int): PublicKey = PublicKey(ByteArray(32) { byte.toByte() })

    private fun u64LE(value: Long): ByteArray {
        val out = ByteArray(8)
        for (i in 0 until 8) out[i] = ((value ushr (8 * i)) and 0xff).toByte()
        return out
    }

    private fun u32LE(value: Int): ByteArray {
        val out = ByteArray(4)
        for (i in 0 until 4) out[i] = ((value ushr (8 * i)) and 0xff).toByte()
        return out
    }

    private fun u16LE(value: Int): ByteArray =
        byteArrayOf((value and 0xff).toByte(), ((value ushr 8) and 0xff).toByte())

    // ── Voucher message bytes ──

    @Test
    fun voucherMessageIsProgramBorshLayout() {
        val bytes = PaymentChannels.voucherMessageBytes(pk(9), 42, 1234)
        assertEquals(48, bytes.size)
        assertContentEquals(pk(9).bytes, bytes.copyOfRange(0, 32))
        assertContentEquals(u64LE(42), bytes.copyOfRange(32, 40))
        assertContentEquals(u64LE(1234), bytes.copyOfRange(40, 48))
    }

    @Test
    fun voucherMessageKnownValues() {
        val bytes = PaymentChannels.voucherMessageBytes(pk(3), 1000, 42)
        assertEquals(48, bytes.size)
        assertContentEquals(u64LE(1000), bytes.copyOfRange(32, 40))
        assertContentEquals(u64LE(42), bytes.copyOfRange(40, 48))
    }

    @Test
    fun voucherMessageDiffersByCumulative() {
        val a = PaymentChannels.voucherMessageBytes(pk(6), 100, 42)
        val b = PaymentChannels.voucherMessageBytes(pk(6), 200, 42)
        assertTrue(!a.contentEquals(b))
    }

    // ── Distribution hash ──

    @Test
    fun distributionHashMatchesProgramPreimageShape() {
        val recipients = listOf(
            PaymentChannels.Distribution(pk(1), 7_500),
            PaymentChannels.Distribution(pk(2), 2_500),
        )
        val hasher = Blake3()
        hasher.update(u32LE(2))
        hasher.update(pk(1).bytes)
        hasher.update(u16LE(7_500))
        hasher.update(pk(2).bytes)
        hasher.update(u16LE(2_500))
        assertContentEquals(hasher.finalize(), PaymentChannels.distributionHash(recipients))
    }

    @Test
    fun distributionHashEmpty() {
        val hasher = Blake3()
        hasher.update(byteArrayOf(0, 0, 0, 0))
        assertContentEquals(hasher.finalize(), PaymentChannels.distributionHash(emptyList()))
    }

    // ── Channel PDA ──

    @Test
    fun channelPdaIsStable() {
        val programId = PaymentChannels.defaultProgramId()
        val (channel1, bump1) = PaymentChannels.findChannelPda(pk(1), pk(2), pk(3), pk(4), 99, programId)
        val (channel2, bump2) = PaymentChannels.findChannelPda(pk(1), pk(2), pk(3), pk(4), 99, programId)
        assertEquals(channel1, channel2)
        assertEquals(bump1, bump2)
    }

    @Test
    fun channelPdaDiffersBySalt() {
        val programId = PaymentChannels.defaultProgramId()
        val (a, _) = PaymentChannels.findChannelPda(pk(1), pk(2), pk(3), pk(4), 1, programId)
        val (b, _) = PaymentChannels.findChannelPda(pk(1), pk(2), pk(3), pk(4), 2, programId)
        assertTrue(a != b)
    }

    // ── Open instruction shape ──

    @Test
    fun buildOpenInstructionShape() {
        val programId = PaymentChannels.defaultProgramId()
        val params = PaymentChannels.OpenChannelParams(
            payer = pk(1),
            payee = pk(2),
            mint = pk(3),
            authorizedSigner = pk(4),
            salt = 7,
            deposit = 1_000_000,
            gracePeriod = 900,
            recipients = emptyList(),
            tokenProgram = PublicKey.fromBase58(Programs.TOKEN_PROGRAM),
            programId = programId,
        )
        val ix = PaymentChannels.buildOpenInstruction(params)
        assertEquals(programId.toBase58(), ix.programId)
        assertEquals(13, ix.accounts.size)
        // Discriminator 1, then OpenArgs Borsh: salt(8) deposit(8) grace(4) recipients-len(4).
        assertEquals(1.toByte(), ix.data[0])
        assertContentEquals(u64LE(7), ix.data.copyOfRange(1, 9))
        assertContentEquals(u64LE(1_000_000), ix.data.copyOfRange(9, 17))
        assertContentEquals(u32LE(900), ix.data.copyOfRange(17, 21))
        assertContentEquals(u32LE(0), ix.data.copyOfRange(21, 25))
        assertEquals(25, ix.data.size)
        // payer is signer + writable, channel/payerATA/channelATA writable.
        assertTrue(ix.accounts[0].isSigner && ix.accounts[0].isWritable)
        assertTrue(ix.accounts[4].isWritable)
    }

    @Test
    fun buildOpenInstructionWithRecipients() {
        val programId = PaymentChannels.defaultProgramId()
        val params = PaymentChannels.OpenChannelParams(
            payer = pk(1),
            payee = pk(2),
            mint = pk(3),
            authorizedSigner = pk(4),
            salt = 0,
            deposit = 500,
            gracePeriod = 0,
            recipients = listOf(PaymentChannels.Distribution(pk(5), 1000)),
            tokenProgram = PublicKey.fromBase58(Programs.TOKEN_PROGRAM),
            programId = programId,
        )
        val ix = PaymentChannels.buildOpenInstruction(params)
        // 1 + 8 + 8 + 4 + 4 + (32 + 2) = 59.
        assertEquals(59, ix.data.size)
        assertContentEquals(u32LE(1), ix.data.copyOfRange(21, 25))
        assertContentEquals(pk(5).bytes, ix.data.copyOfRange(25, 57))
        assertContentEquals(u16LE(1000), ix.data.copyOfRange(57, 59))
    }

    @Test
    fun buildTopUpInstructionShape() {
        val programId = PaymentChannels.defaultProgramId()
        val ix = PaymentChannels.buildTopUpInstruction(
            payer = pk(1),
            channel = pk(2),
            mint = pk(3),
            amount = 250_000,
            tokenProgram = PublicKey.fromBase58(Programs.TOKEN_PROGRAM),
            programId = programId,
        )
        assertEquals(programId.toBase58(), ix.programId)
        assertEquals(6, ix.accounts.size)
        assertEquals(3.toByte(), ix.data[0])
        assertContentEquals(u64LE(250_000), ix.data.copyOfRange(1, 9))
        assertEquals(9, ix.data.size)
    }

    // ── Ed25519 verify instruction ──

    @Test
    fun ed25519VerifyInstructionLayout() {
        val signer = pk(8)
        val signature = ByteArray(64) { 1 }
        val message = PaymentChannels.voucherMessageBytes(pk(9), 42, 1234)
        val ix = PaymentChannels.buildEd25519VerifyInstruction(signer, signature, message)
        assertEquals(PaymentChannels.ED25519_PROGRAM_ID, ix.programId)
        // header(16) + pubkey(32) + sig(64) + message(48) = 160.
        assertEquals(16 + 32 + 64 + message.size, ix.data.size)
        assertEquals(1.toByte(), ix.data[0])
        assertEquals(0.toByte(), ix.data[1])
        // public key offset 16, signature offset 48, message offset 112.
        assertContentEquals(u16LE(48), ix.data.copyOfRange(2, 4))
        assertContentEquals(pk(8).bytes, ix.data.copyOfRange(16, 48))
        assertContentEquals(signature, ix.data.copyOfRange(48, 112))
        assertContentEquals(message, ix.data.copyOfRange(112, 160))
    }
}

File: go/protocols/mpp/program/payment_channels_test.go
package program

import (
	"encoding/binary"
	"testing"

	solana "github.com/gagliardetto/solana-go"
	"lukechampine.com/blake3"
)

func pk(b byte) solana.PublicKey {
	var out solana.PublicKey
	for i := range out {
		out[i] = b
	}
	return out
}

// Golden vector: mirrors voucher_message_is_program_borsh_layout in the Rust
// spine. channel=pk(9), cumulative=42, expires=1234.
func TestVoucherMessageBytesGoldenLayout(t *testing.T) {
	channel := pk(9)
	bytes := VoucherMessageBytes(channel, 42, 1234)
	if len(bytes) != VoucherMessageLen {
		t.Fatalf("voucher message length = %d, want %d", len(bytes), VoucherMessageLen)
	}
	if got := bytes[:32]; string(got) != string(channel[:]) {
		t.Fatalf("voucher channel prefix mismatch")
	}
	if got := binary.LittleEndian.Uint64(bytes[32:40]); got != 42 {
		t.Fatalf("voucher cumulative LE = %d, want 42", got)
	}
	if got := int64(binary.LittleEndian.Uint64(bytes[40:48])); got != 1234 {
		t.Fatalf("voucher expires LE = %d, want 1234", got)
	}
}

// Golden vector: mirrors voucher_data_message_bytes_with_nonce (channel=pk(3),
// cumulative=1000, expires=42).
func TestVoucherMessageBytesGoldenWithKnownValues(t *testing.T) {
	bytes := VoucherMessageBytes(pk(3), 1000, 42)
	if binary.LittleEndian.Uint64(bytes[32:40]) != 1000 {
		t.Fatalf("cumulative mismatch")
	}
	if int64(binary.LittleEndian.Uint64(bytes[40:48])) != 42 {
		t.Fatalf("expires mismatch")
	}
}

// Golden vector: mirrors distribution_hash_matches_program_preimage_shape in
// the Rust spine. The preimage is count(u32 LE) || (recipient(32)||bps(u16 LE)).
func TestDistributionHashGolden(t *testing.T) {
	recipients := []Distribution{
		{Recipient: pk(1), Bps: 7500},
		{Recipient: pk(2), Bps: 2500},
	}
	hasher := blake3.New(32, nil)
	count := make([]byte, 4)
	binary.LittleEndian.PutUint32(count, 2)
	_, _ = hasher.Write(count)
	r1 := pk(1)
	_, _ = hasher.Write(r1[:])
	bps := make([]byte, 2)
	binary.LittleEndian.PutUint16(bps, 7500)
	_, _ = hasher.Write(bps)
	r2 := pk(2)
	_, _ = hasher.Write(r2[:])
	binary.LittleEndian.PutUint16(bps, 2500)
	_, _ = hasher.Write(bps)
	var want [32]byte
	copy(want[:], hasher.Sum(nil))

	got := DistributionHash(recipients)
	if got != want {
		t.Fatalf("distribution hash mismatch\n got=%x\nwant=%x", got, want)
	}
}

func TestDistributionHashEmpty(t *testing.T) {
	got := DistributionHash(nil)
	hasher := blake3.New(32, nil)
	_, _ = hasher.Write([]byte{0, 0, 0, 0})
	var want [32]byte
	copy(want[:], hasher.Sum(nil))
	if got != want {
		t.Fatalf("empty distribution hash mismatch")
	}
}

// Golden vector: mirrors channel_pda_is_stable in the Rust spine. The derived
// PDA must equal CreateProgramAddress with the discovered bump.
func TestChannelPDAStable(t *testing.T) {
	programID := DefaultProgramID()
	channel, bump, err := FindChannelPDA(pk(1), pk(2), pk(3), pk(4), 99, programID)
	if err != nil {
		t.Fatal(err)
	}
	saltLE := make([]byte, 8)
	binary.LittleEndian.PutUint64(saltLE, 99)
	expected, err := solana.CreateProgramAddress([][]byte{
		[]byte(ChannelSeed),
		pk(1).Bytes(), pk(2).Bytes(), pk(3).Bytes(), pk(4).Bytes(),
		saltLE, {bump},
	}, programID)
	if err != nil {
		t.Fatal(err)
	}
	if !channel.Equals(expected) {
		t.Fatalf("channel PDA mismatch: %s != %s", channel, expected)
	}
}

func TestBuildOpenInstructionShape(t *testing.T) {
	params := OpenChannelParams{
		Payer:            pk(1),
		Payee:            pk(2),
		Mint:             pk(3),
		AuthorizedSigner: pk(4),
		Salt:             7,
		Deposit:          1_000_000,
		GracePeriod:      900,
		TokenProgram:     solana.TokenProgramID,
		ProgramID:        DefaultProgramID(),
	}
	ix, err := BuildOpenInstruction(params)
	if err != nil {
		t.Fatal(err)
	}
	if !ix.ProgramID().Equals(DefaultProgramID()) {
		t.Fatalf("open instruction program id mismatch")
	}
	if len(ix.Accounts()) != 13 {
		t.Fatalf("open instruction accounts = %d, want 13", len(ix.Accounts()))
	}
	data, err := ix.Data()
	if err != nil {
		t.Fatal(err)
	}
	if data[0] != openDiscriminator {
		t.Fatalf("open discriminator = %d, want %d", data[0], openDiscriminator)
	}
	// discriminator(1) + salt(8) + deposit(8) + grace(4) + recipients len(4) = 25
	if len(data) != 25 {
		t.Fatalf("open data len = %d, want 25", len(data))
	}
	if binary.LittleEndian.Uint64(data[1:9]) != 7 {
		t.Fatalf("open salt mismatch")
	}
	if binary.LittleEndian.Uint64(data[9:17]) != 1_000_000 {
		t.Fatalf("open deposit mismatch")
	}
	if binary.LittleEndian.Uint32(data[17:21]) != 900 {
		t.Fatalf("open grace mismatch")
	}
}

func TestBuildTopUpInstructionShape(t *testing.T) {
	ix, err := BuildTopUpInstruction(pk(1), pk(5), pk(3), 42, solana.TokenProgramID, DefaultProgramID())
	if err != nil {
		t.Fatal(err)
	}
	if len(ix.Accounts()) != 6 {
		t.Fatalf("topup accounts = %d, want 6", len(ix.Accounts()))
	}
	data, err := ix.Data()
	if err != nil {
		t.Fatal(err)
	}
	if data[0] != topUpDiscriminator {
		t.Fatalf("topup discriminator = %d, want %d", data[0], topUpDiscriminator)
	}
	if binary.LittleEndian.Uint64(data[1:9]) != 42 {
		t.Fatalf("topup amount mismatch")
	}
}

func TestBuildEd25519VerifyInstructionLayout(t *testing.T) {
	var sig [64]byte
	for i := range sig {
		sig[i] = byte(i)
	}
	message := VoucherMessageBytes(pk(9), 42, 1234)
	ix, err := BuildEd25519VerifyInstruction(pk(7), sig, message)
	if err != nil {
		t.Fatal(err)
	}
	if !ix.ProgramID().Equals(solana.MustPublicKeyFromBase58(Ed25519ProgramID)) {
		t.Fatalf("ed25519 program id mismatch")
	}
	data, err := ix.Data()
	if err != nil {
		t.Fatal(err)
	}
	// header: count=1, pad=0, then 7 u16 fields = 2 + 14 = 16
	if data[0] != 1 || data[1] != 0 {
		t.Fatalf("ed25519 header count/pad mismatch")
	}
	signer := pk(7)
	if string(data[16:48]) != string(signer[:]) {
		t.Fatalf("ed25519 public key offset mismatch")
	}
	if string(data[48:112]) != string(sig[:]) {
		t.Fatalf("ed25519 signature offset mismatch")
	}
	if string(data[112:]) != string(message) {
		t.Fatalf("ed25519 message offset mismatch")
	}
}
